### PR TITLE
fix: Simplified `IFileSystem`

### PR DESCRIPTION
**Pull request by bitsandfoxes**
## TLDR;

The responsibility to check whether a system is allowed to write to disk or not falls onto the system itself. The `ReadOnlyFileSystem` is only there to prevent regressions and crashes on platforms like the Switch.

Any attempts to write to disk and fail can now simply logged as `Error` because that's what they are.

## Context

The idea was to have `IFileSystem` be the abstraction layer and be swapped between `ReadOnly` and `ReadWrite` and have it be clever enough to log whether the operation failed or was disallowed.

Having it smart enough is not really feasable. The caller is still interested whether it failed or was disallowed for logging purposes. Having the caller simply attempt to write to disk and then not care whether it was allowed to do so is not possible.

So we end up checking the success/failure/disallowed with every call. That is noisy. So we're checking whether we're allowed to write to disk before even attempting. This making the `FileOperationResult` pointless. We can just log and skip when we know we can't write to disk in the first place.

### CODE REVIEW

File: src/Sentry/ISentryJsonSerializable.cs [hunk 1]
```diff
@@ -17,21 +17,3 @@ public interface ISentryJsonSerializable
...
-            return;
-        }
-
-        using var writer = new Utf8JsonWriter(file);
-
-        serializable.WriteTo(writer, logger);
-        writer.Flush();
-        file.Dispose();
-    }
-}
```
Comment (bitsandfoxes) [inline src/Sentry/ISentryJsonSerializable.cs:37]: This is only used once in the `GlobalSessionsManager`. Success/Failure is getting handled there and does not need to be abstracted away into an extension.

**Comment by bruno-garcia** [comment]
Follow up to:
* https://github.com/getsentry/sentry-dotnet/pull/3614

File: src/Sentry/GlobalSessionManager.cs [hunk 2]
```diff
@@ -91,6 +104,12 @@ private void DeletePersistedSession()
             return;
         }
 
+        if (_options.DisableFileWrite)
+        {
+            _options.LogInfo("File write has been disabled via the options. Skipping deleting persisted session.");
```
Comment (bruno-garcia) [inline src/Sentry/GlobalSessionManager.cs:109]: "skipping deletion if session files"

when we're reading the debug level log we're not sure how individual messages relate so could be helpful to be explicit

File: src/Sentry/Internal/Http/CachingTransport.cs [hunk 3]
```diff
@@ -78,6 +78,13 @@ private CachingTransport(ITransport innerTransport, SentryOptions options, bool
             options.TryGetProcessSpecificCacheDirectoryPath() ??
             throw new InvalidOperationException("Cache directory or DSN is not set.");
 
+        // Sanity check: This should never happen in the first place.
+        // We check for `DisableFileWrite` before creating the CachingTransport.
+        if (_options.DisableFileWrite)
```
Comment (bruno-garcia) [inline src/Sentry/Internal/Http/CachingTransport.cs:83]: if this never happens consider Debug.Assert. 

Adding `throw` has side effects such as the method can no longer be inlined.

File: src/Sentry/Internal/IFileSystem.cs [hunk 4]
```diff
@@ -1,14 +1,12 @@
...
-{
-    Success,
-    Failure,
-    Disabled
-}
-
 internal interface IFileSystem
 {
+    // Note: You are responsible for handling success/failure when attempting to write to disk.
+    // You are required to check for `Options.FileWriteDisabled` whether you are allowed to call any writing operations.
```
Comment (bruno-garcia) [inline src/Sentry/Internal/IFileSystem.cs:6]: This is a good call out. Adding a sexton like <example> or <note> (can't recall the options) above methods in the interface with a snippet might make it even more obvious

File: src/Sentry/Internal/ReadOnlyFilesystem.cs [hunk 5]
```diff
@@ -2,20 +2,24 @@ namespace Sentry.Internal;
 
 internal class ReadOnlyFileSystem : FileSystemBase
 {
-    public override FileOperationResult CreateDirectory(string path) => FileOperationResult.Disabled;
+    // Note: You are responsible for handling success/failure when attempting to write to disk.
```
Comment (bruno-garcia) [inline src/Sentry/Internal/ReadOnlyFilesystem.cs:5]: this won't show up in the IDE at call site. Needs <summary>

**Review by bruno-garcia — APPROVED** [review]
Having integration tests related to e2e and with file I/O (sessions etc) have two runs, one with this option set to true and other to false, would really help us avoid regression

File: src/Sentry/Internal/ReadOnlyFilesystem.cs (hunk 5 above)
Comment (jamescrosswell) [reply src/Sentry/Internal/ReadOnlyFilesystem.cs:5]: These interfaces and classes are all internal, so it looks like the only place to document this is in some remarks or something on `SentryOptions.DisableFileWrite`.

**Comment by jamescrosswell** [comment]
> Having the caller simply attempt to write to disk and then not care whether it was allowed to do so is not possible.

@bitsandfoxes can you provide a bit more context around when/why that's not possible?

**Comment by bitsandfoxes** [comment]
> @bitsandfoxes can you provide a bit more context around when/why that's not possible?

Because the caller typically would like to log whether the failed write attempt failed because of a failure or because writing has been disabled. So I ended up with constructs like this
```
var result = _options.FileSystem.CreateDirectory(_persistenceDirectoryPath);
if (result is not FileOperationResult.Success)
{
    if (result is not FileOperationResult.Disabled)
    {
        _options.LogDebug("Disabled.");
    }
    else
    {
        _options.LogError("Failed.");    
    }
    
    return;
}
```
all over the place. So I could not escape checking for `Options.FileWriteDisabled`, either directly or indirectly anyway. I think it's simpler to be explicit about it.
